### PR TITLE
reduction on complex numbers, added volatile copy and assignment

### LINKED ISSUE #2369: nvFuser doesn't support complex reductions
```
E   CUDA NVRTC compile error: __tmp_kernel5.cu(3993): error: no operator "=" matches these operands
E               operand types are: volatile CudaCodeGen::std::complex<double> = CudaCodeGen::std::complex<double>
E             detected during instantiation of "void CudaCodeGen::reduction::gridReduce<X_BLOCK,Y_BLOCK,Z_BLOCK,X_THREAD,Y_THREAD,Z_THREAD,PERSISTENT_REDUCTION,T,Func>(T &, const T &, Func, volatile T *, CudaCodeGen::int64_t *, T *, __nv_bool, __nv_bool, T, CudaCodeGen::nvfuser_index_t, CudaCodeGen::nvfuser_index_t) [with X_BLOCK=true, Y_BLOCK=false, Z_BLOCK=false, X_THREAD=true, Y_THREAD=false, Z_THREAD=false, PERSISTENT_REDUCTION=false, T=CudaCodeGen::std::complex<double>, Func=lambda [](CudaCodeGen::std::complex<double> &, CudaCodeGen::std::complex<double>)->void]" 
E   (7913): here
E   
E   __tmp_kernel5.cu(3848): error: class "CudaCodeGen::std::complex<double>" has no suitable copy constructor
E             detected during:
E               instantiation of "void CudaCodeGen::reduction::gridReduceLastBlock<X_THREAD,Y_THREAD,Z_THREAD,T,Func>(T &, const volatile T *, CudaCodeGen::nvfuser_index_t, CudaCodeGen::nvfuser_index_t, Func, T *, __nv_bool, T) [with X_THREAD=false, Y_THREAD=true, Z_THREAD=true, T=CudaCodeGen::std::complex<double>, Func=lambda [](CudaCodeGen::std::complex<double> &, CudaCodeGen::std::complex<double>)->void]" 
E   (4019): here
E               instantiation of "void CudaCodeGen::reduction::gridReduce<X_BLOCK,Y_BLOCK,Z_BLOCK,X_THREAD,Y_THREAD,Z_THREAD,PERSISTENT_REDUCTION,T,Func>(T &, const T &, Func, volatile T *, CudaCodeGen::int64_t *, T *, __nv_bool, __nv_bool, T, CudaCodeGen::nvfuser_index_t, CudaCodeGen::nvfuser_index_t) [with X_BLOCK=true, Y_BLOCK=false, Z_BLOCK=false, X_THREAD=true, Y_THREAD=false, Z_THREAD=false, PERSISTENT_REDUCTION=false, T=CudaCodeGen::std::complex<double>, Func=lambda [](CudaCodeGen::std::complex<double> &, CudaCodeGen::std::complex<double>)->void]" 
E   (7913): here
E   
E   2 errors detected in the compilation of "__tmp_kernel5.cu".
```

### PR DESCRIPTION
**Pull request by liqiangxl**
copy constructor must be declared within class. So I need to directly patch `aten/src/ATen/cuda/llvm_complex.cpp` to allow the use of volatile in reduction.

Fixes #2369

### CODE REVIEW

**Comment by naoyam** [comment]
While not ideal, it should be possible to make the reductions work without modifying the LLVM header file. For serial reductions, they are dynamically generated, so we could tweak our codegen to emit different code for volatile complex types. For block/grid reductions, we are using device functions provided as part of the nvfuser "runtime". For those cases, we could provide a copy functor to the device functions so that the overload cases for volatile types aren't necessary. 

It's certainly easier from the codegen side if we could just modify the header file as done in the PR. However, if we don't want to   make a modification of the external code, there's an alternative option. Another option would be inserting the overload definitions programmatically into the strings returned from `at::cuda::get_complex_body_string()`, etc.

**Comment by liqiangxl** [comment]
`aten/src/ATen/cuda/llvm_complex.cpp` in Pytorch is copy-pasted (with modification) from the original llvm file. If we don't want to modify that file, can we just make a copy-past to our own runtime folder and combine everything we added to that file for the support of complex number?

File: third_party/nvfuser/runtime/complex_number.cu [hunk 1]
```diff
@@ -0,0 +1,1485 @@
...
+class __promote_imp<_A1, void, void, true> {
+ public:
+  typedef typename __numeric_type<_A1>::type type;
+  static const bool value = true;
+};
+
+template <class _A1, class _A2 = void, class _A3 = void>
+class __promote : public __promote_imp<_A1, _A2, _A3> {};
+
+} // namespace std
```
Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/complex_number.cu]: Could you move this part to a separate runtime file?

**Review by zasdfgbnm — COMMENTED** [review]
Could you skim through our tests in `third_party/nvfuser/test` to find the tests that check reductions with different dtypes, and add complex to the dtype list? I do this by searching `DataType::Double` and read all matches manually to determine if a test is interesting to me. Here is a list of tests I found: `FusionGroupAllreduce5_CUDA`, `FusionReductionSchedulerNoODimShmoo_CUDA`, `FusionReductionSchedulerDimShmoo_CUDA`, `FusionWelfordShmoo_CUDA`

**Comment by csarofeen** [comment]
Is this ready for another review?

File: third_party/nvfuser/csrc/codegen.cpp [hunk 2]
```diff
@@ -1913,8 +1914,14 @@ class CudaKernelGenerator : private OptOutConstDispatch {
             .append("[")
             .append(work_buffer_offset)
             .append("]");
-        init_vals.arg(genInline(grouped_grop->initVal(expr_index)));
-
+        auto iv = (grouped_grop->initVal(expr_index));
+        // TODO: why data type doesn't match? e.g. complex<float> <--->
+        // complex<double>
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/codegen.cpp]: Python scalar only has double, int64_t and complex double, there is no float, int32 and complex float. PyTorch scalar has the same design as python scalar, so the dtype might be mismatch here. I am fine with the change below, but please update this comment.

File: third_party/nvfuser/csrc/lower_utils.cpp [hunk 3]
```diff
@@ -249,6 +249,13 @@ c10::optional<IterDomain*> getMaybeWarpReductionDim(
   }
 
   auto tv_in = getTv(input);
+
+  // __shfl_xor_sync() doesn't support complex number
+  if (tv_in->dtype() == DataType::ComplexFloat ||
+      tv_in->dtype() == DataType::ComplexDouble) {
+    return c10::nullopt;
+  }
+
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/lower_utils.cpp:258]: It is fine to do so in this PR, but in the future, we should consider using `__shfl_xor_sync` to exchange the real and imag part of the complex separately so we can do warp reduction on complex numbers.

File: third_party/nvfuser/runtime/basic_type_traits.cu [hunk 4]
```diff
@@ -0,0 +1,172 @@
+namespace std {
```
Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/basic_type_traits.cu:17]: Should we protect this file with
```C++
#ifdef __NVCC__
#include <type_traits>
#else
....
#endif
```
Please try a `PYTORCH_NVFUSER_DUMP="cuda_to_file"` run on something, and use nvcc to compile the dumped file to check if it compiles.

Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/basic_type_traits.cu:17]: Should we put a copyright section and a pointer to the original code here for LLVM copied code?

File: third_party/nvfuser/runtime/complex_number.cu [hunk 5]
```diff
@@ -0,0 +1,1313 @@
+#ifndef __NVCC__
+#define POS_INFINITY __int_as_float(0x7f800000)
+#define INFINITY POS_INFINITY
+#define NEG_INFINITY __int_as_float(0xff800000)
+#define NAN __int_as_float(0x7fffffff)
+
+namespace std {
```
Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/complex_number.cu:20]: Same here, should we put a copyright section and a pointer to the original code here for LLVM copied code?

File: third_party/nvfuser/test/test_gpu2.cpp [hunk 6]
```diff
@@ -2873,6 +2873,10 @@ TEST_F(NVFuserTest, FusionWelfordShmoo_CUDA) {
   //   Detected abs error of: 3.8062
   //     absolute tolerance was set to 2.23704e-06
   //     and relative tolerance set to 2.23704e-08
+  // Reason: variance of complex numbers is a real value instead of a complex
+  // number to enable complex number with Welford, we need to either (1)
+  // find/invent a specific version of Welford for complex numbers or (2)
+  // translate Welford to two-pass approach
```
Comment (zasdfgbnm) [inline third_party/nvfuser/test/test_gpu2.cpp:2879]: Thanks for the investigation.

File: third_party/nvfuser/runtime/helpers.cu [hunk 7]
```diff
@@ -61,30 +61,6 @@ __device__ float lerp(float start, float end, float weight) {
   }
 }
 
-#ifndef __NVCC__
```
Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/helpers.cu:64]: Nice cleanup!

File: third_party/nvfuser/test/test_gpu_fused_reduction.cpp [hunk 8]
```diff
@@ -1219,12 +1219,18 @@ TEST_F(NVFuserTest, FusionGroupAllreduce5_CUDA) {
...
-      add(castOp(DataType::Double, tv3), tv7), castOp(DataType::Double, tv11));
+  auto tv12 = makeSymbolicTensor(1, DataType::ComplexFloat);
+  fusion.addInput(tv12);
+  auto tv13 = sum(tv12, {0});
+  auto tv14 = broadcast(tv13, {true});
+  auto tv15 = div(tv12, tv14);
 
+  auto outFDI = add(
+      add(castOp(DataType::Double, tv3), tv7), castOp(DataType::Double, tv11));
+  auto out = add(outFDI, castOp(DataType::Double, tv15));
```
Comment (zasdfgbnm) [inline third_party/nvfuser/test/test_gpu_fused_reduction.cpp]: Should we cast to complex double here?

File: third_party/nvfuser/csrc/codegen.cpp (hunk 2 above)
Comment (liqiangxl) [reply third_party/nvfuser/csrc/codegen.cpp]: done. thanks for the background knowledge!

File: third_party/nvfuser/csrc/lower_utils.cpp (hunk 3 above)
Comment (liqiangxl) [reply third_party/nvfuser/csrc/lower_utils.cpp:258]: issue #2526 is added to to track this.

File: third_party/nvfuser/runtime/basic_type_traits.cu (hunk 4 above)
Comment (liqiangxl) [reply third_party/nvfuser/runtime/basic_type_traits.cu:17]: nice catch! we need this protection. However, the generated code still can't compile with nvcc due to the lack of non-volatile to volatile copy of complex numbers in `<complex>`. error msg:
```
__tmp_kernel1.cu(3426): error: no operator "=" matches these operands
            operand types are: volatile std::complex<float> = const std::complex<float>
          detected during:
            instantiation of "void CudaCodeGen::TupleCopy<DstType, SrcType, num_vals>::copy(DstType &, CudaCodeGen::nvfuser_index_t, const SrcType &, CudaCodeGen::nvfuser_index_t) [with DstType=CudaCodeGen::PtrTupleBase<true, float, double, CudaCodeGen::int64_t, std::complex<float>>, SrcType=CudaCodeGen::LocalTuple<float, double, CudaCodeGen::int64_t, std::complex<float>>, num_vals=4]" 
```

Comment (liqiangxl) [reply third_party/nvfuser/runtime/basic_type_traits.cu:17]: added

File: third_party/nvfuser/runtime/complex_number.cu (hunk 5 above)
Comment (liqiangxl) [reply third_party/nvfuser/runtime/complex_number.cu:20]: added

File: third_party/nvfuser/test/test_gpu2.cpp (hunk 6 above)
Comment (liqiangxl) [reply third_party/nvfuser/test/test_gpu2.cpp:2879]: issue #2527 is added to track this.

File: third_party/nvfuser/test/test_gpu_fused_reduction.cpp (hunk 8 above)
Comment (liqiangxl) [reply third_party/nvfuser/test/test_gpu_fused_reduction.cpp]: this is just pick the real part and cast to double. All results from other data types are casted to double, so does the reference results from torch.

File: third_party/nvfuser/runtime/basic_type_traits.cu (hunk 4 above)
Comment (zasdfgbnm) [reply third_party/nvfuser/runtime/basic_type_traits.cu:17]: Are you saying that, if the kernel is complex, then we are still unable to compile with nvcc. But if it is not complex then it is working? If so, then it is fine for now. I believe @mmigdal-nv is using nvcc with matmul quite often, and we need to make sure this PR doesn't break non-complex matmul use case.

File: third_party/nvfuser/test/test_gpu_fused_reduction.cpp (hunk 8 above)
Comment (zasdfgbnm) [reply third_party/nvfuser/test/test_gpu_fused_reduction.cpp]: Should we cast all results from other data types and reference results to complex double as well? Casting to double discard imag so we are not checking the correctness of imag.

File: third_party/nvfuser/runtime/basic_type_traits.cu (hunk 4 above)
Comment (liqiangxl) [reply third_party/nvfuser/runtime/basic_type_traits.cu:17]: It's only influence reduction of complex numbers.

File: third_party/nvfuser/test/test_gpu_fused_reduction.cpp (hunk 8 above)
Comment (liqiangxl) [reply third_party/nvfuser/test/test_gpu_fused_reduction.cpp]: updated this test with all results converted to complex<double>

**Review by zasdfgbnm — APPROVED** [review]
LGTM, thanks for adding reduction support for complex numbers!

**Pull request merged**